pkg/cluster/internal/create/actions/kubeadmjoin/join.go

```diff
@@ -100,10 +100,27 @@ func joinWorkers(
 	defer ctx.Status.End(false)
 
 	// TODO(bentheelder): this should be concurrent
+	errChan := make(chan error, len(workers))
+	defer close(errChan)
 	for _, node := range workers {
-		if err := runKubeadmJoin(ctx, allNodes, &node); err != nil {
+		node := node // capture loop variable
+		go func() {
+			errChan <- runKubeadmJoin(ctx, allNodes, &node)
+		}()
+	}
+
+	// watch for all worker joins to finish
+	// NOTE: we don't use a waitgroup because we want to exit early
+	count := 0
+	for err := range errChan {
+		// if any errored, bail out
+		if err != nil {
 			return err
 		}
+		count++
+		if count == len(workers) {
```

Comment:
Does ctx already have a timeout? Do we need to also wait for `ctx.Done` using a switch?

Reply:
it does not yet. 
I don't think we care though, we'll list up the containers and forcibly delete them on exit if retain is not set, if it is set we wouldn't have done anything either way, we'll just leave them around to poke at for debugging